man/vmemctrl.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VMEMCTRL "8" "May 2013" "vmemctrl  --" "System Administration"
.SH NAME
vmemctrl \- Control the OOM bias for vservers
.SH DESCRIPTION
.SS "Usage:"
.IP
vmemctrl (\fB\-\-set\fR|\-\-get) [\-\-xid <xid>] [\-\-badness <OOM bias>]
.IP
[\-\-] [<command> <args>*]
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2007 Daniel Hokka Zakrisson
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vmemctrl.c
// $Id$

// Copyright (C) 2007 Daniel Hokka Zakrisson
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "vserver.h"
#include "util.h"

#include <lib/internal.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>

#define ENSC_WRAPPERS_PREFIX	"vmemctrl: "
#define ENSC_WRAPPERS_VSERVER	1
#define ENSC_WRAPPERS_UNISTD	1
#include "wrappers.h"

#define CMD_HELP	0x1000
#define CMD_VERSION	0x1001

#define CMD_SET		0x2000
#define CMD_GET		0x2001

#define CMD_XID		0x4000
#define CMD_BADNESS	0x4001

int wrapper_exit_code = 255;


static struct option const
CMDLINE_OPTIONS[] = {
  { "help",     no_argument,        0, CMD_HELP },
  { "version",  no_argument,        0, CMD_VERSION },
  { "set",      no_argument,        0, CMD_SET },
  { "get",      no_argument,        0, CMD_GET },
  { "xid",      required_argument,  0, CMD_XID },
  { "badness",  required_argument,  0, CMD_BADNESS },
  { 0,0,0,0 }
};

struct Arguments {
  xid_t		xid;
  int64_t	badness;
  bool		do_set;
  bool		do_get;
  uint32_t	mask;
};

static inline int
hasArg(struct Arguments *args, uint32_t arg)
{
  return (args->mask & arg) == arg;
}

static void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:\n  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " (--set|--get) [--xid <xid>] [--badness <OOM bias>]\n"
	    "        [--] [<command> <args>*]\n\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vmemctrl " VERSION " -- \n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2007 Daniel Hokka Zakrisson\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static inline void
doset(struct Arguments *args)
{
  if (hasArg(args, CMD_BADNESS) &&
      vc_set_badness(args->xid, args->badness) == -1) {
    perror(ENSC_WRAPPERS_PREFIX "vc_set_badness()");
    exit(wrapper_exit_code);
  }
}

static inline void
doget(struct Arguments *args)
{
  int64_t badness;
  char buf[32];
  size_t l;
  if (vc_get_badness(args->xid, &badness) == -1) {
    perror(ENSC_WRAPPERS_PREFIX "vc_get_badness()");
    exit(wrapper_exit_code);
  }
  l = utilvserver_fmt_int64(buf, badness);
  buf[l] = '\0';
  WRITE_STR(1, buf);
  WRITE_MSG(1, "\n");
}

int main (int argc, char *argv[])
{
  struct Arguments args = {
    .do_set	= false,
    .do_get	= false,
    .xid	= VC_NOCTX,
    .badness	= 0,
    .mask	= 0,
  };
  
  while (1) {
    int		c = getopt_long(argc, argv, "+", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0], 0);
      case CMD_VERSION	:  showVersion();
      case CMD_XID	:  args.xid       = Evc_xidopt2xid(optarg,true); break;
      case CMD_SET	:  args.do_set    = true; break;
      case CMD_GET	:  args.do_get    = true; break;
      case CMD_BADNESS	: {
	char *endptr;
	args.badness = strtoll(optarg, &endptr, 0);
	if (*endptr) {
	  WRITE_MSG(2, ENSC_WRAPPERS_PREFIX "Badness '");
	  WRITE_STR(2, optarg);
	  WRITE_MSG(2, "' is not an integer\n");
	  exit(wrapper_exit_code);
	}
	break;
      }
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	exit(wrapper_exit_code);
	break;
    }
    args.mask |= c;
  }

  if (args.xid == VC_NOCTX) args.xid = Evc_get_task_xid(0);

  if (!args.do_set && !args.do_get) {
    WRITE_MSG(2, "No operation specified; try '--help' for more information\n");
    exit(wrapper_exit_code);
  }
  else if (((args.do_set ? 1 : 0) + (args.do_get ? 1 : 0)) > 1) {
    WRITE_MSG(2, "Multiple operations specified; try '--help' for more information\n");
    exit(wrapper_exit_code);
  }

  if (args.do_set)
    doset(&args);
  else if (args.do_get)
    doget(&args);

  if (optind != argc)
    Eexecvp (argv[optind],argv+optind);
  return EXIT_SUCCESS;
}
